gcloud/datastore/helpers.py

```diff
@@ -106,16 +155,67 @@ def entity_from_protobuf(pb):
                                  'be indexed or all excluded from indexes.')
 
             if not indexed_values.pop():
-                exclude_from_indexes.append(property_pb.name)
+                exclude_from_indexes.append(prop_name)
         else:
             if not property_pb.value.indexed:
-                exclude_from_indexes.append(property_pb.name)
+                exclude_from_indexes.append(prop_name)
 
     entity = Entity(key=key, exclude_from_indexes=exclude_from_indexes)
     entity.update(entity_props)
+    entity._meanings.update(entity_meanings)
```

Comment:
@tseaver When testing I was storing the original protobuf response by adding

```python
import copy
entity._orig_pb = copy.deepcopy(pb)
```

before returning `entity`.

Most crucially, I used this to verify that

```python
>>> e = client.get(k)
>>> client.put(e)
>>> e2 = client.get(k)
>>> e._orig_pb == e2._orig_pb
True
```